// frontend/tailwind.config.js
/** @type {import('tailwindcss').Config} */
export default {
  darkMode: 'class',
  content: ['./index.html', './src/**/*.{js,ts,jsx,tsx}'],
  theme: {
    extend: {
      colors: {
        surface: '#faf8ff',
        'surface-dim': '#d9d9e5',
        'surface-bright': '#faf8ff',
        'surface-container-lowest': '#ffffff',
        'surface-container-low': '#f3f3fe',
        'surface-container': '#ededf9',
        'surface-container-high': '#e7e7f3',
        'surface-container-highest': '#e1e2ed',
        'on-surface': '#191b23',
        'on-surface-variant': '#434655',
        'inverse-surface': '#2e3039',
        'inverse-on-surface': '#f0f0fb',
        outline: '#737686',
        'outline-variant': '#c3c6d7',
        'surface-tint': '#0053db',
        primary: {
          DEFAULT: '#004ac6',
          container: '#2563eb',
          fixed: '#dbe1ff',
          'fixed-dim': '#b4c5ff',
          foreground: '#ffffff',
          'on-container': '#eeefff',
        },
        secondary: {
          DEFAULT: '#712ae2',
          container: '#8a4cfc',
          fixed: '#eaddff',
          'fixed-dim': '#d2bbff',
          foreground: '#ffffff',
          'on-container': '#fffbff',
        },
        tertiary: {
          DEFAULT: '#943700',
          container: '#bc4800',
          fixed: '#ffdbcd',
          'fixed-dim': '#ffb596',
          foreground: '#ffffff',
          'on-container': '#ffede6',
        },
        error: {
          DEFAULT: '#ba1a1a',
          container: '#ffdad6',
          foreground: '#ffffff',
          'on-container': '#93000a',
        },
        success: {
          DEFAULT: '#059669',
          container: '#d1fae5',
          foreground: '#ffffff',
        },
        background: '#faf8ff',
        'on-background': '#191b23',
        'surface-variant': '#e1e2ed',
        match: {
          green: '#22c55e',
          blue: '#3b82f6',
          purple: '#8b5cf6',
          orange: '#f97316',
          red: '#ef4444',
        },
      },
      fontFamily: {
        sans: ['Inter', 'system-ui', 'sans-serif'],
        label: ['Geist', 'Inter', 'system-ui', 'sans-serif'],
      },
      fontSize: {
        'headline-xl': ['48px', { lineHeight: '1.1', letterSpacing: '-0.02em', fontWeight: '700' }],
        'headline-lg': ['32px', { lineHeight: '1.2', letterSpacing: '-0.015em', fontWeight: '600' }],
        'headline-md': ['24px', { lineHeight: '1.3', fontWeight: '600' }],
        'body-lg': ['18px', { lineHeight: '1.6', fontWeight: '400' }],
        'body-md': ['16px', { lineHeight: '1.5', fontWeight: '400' }],
        'label-md': ['14px', { lineHeight: '1.4', letterSpacing: '0.01em', fontWeight: '500' }],
        'label-sm': ['12px', { lineHeight: '1.2', fontWeight: '600' }],
      },
      spacing: {
        gutter: '24px',
        unit: '4px',
        'margin-mobile': '16px',
        'margin-desktop': '48px',
      },
      borderRadius: {
        DEFAULT: '0.5rem',
        lg: '0.75rem',
        xl: '1rem',
        '2xl': '1.5rem',
      },
      boxShadow: {
        card: '0 4px 20px rgba(0, 0, 0, 0.04)',
        elevated: '0 12px 40px rgba(0, 0, 0, 0.08)',
        'ai-glow': '0 0 20px rgba(37, 99, 235, 0.12)',
      },
      animation: {
        'fade-in': 'fadeIn 0.2s ease-out',
        'slide-up': 'slideUp 0.3s ease-out',
        'pulse-slow': 'pulse 3s cubic-bezier(0.4, 0, 0.6, 1) infinite',
      },
      keyframes: {
        fadeIn: { from: { opacity: '0' }, to: { opacity: '1' } },
        slideUp: { from: { opacity: '0', transform: 'translateY(8px)' }, to: { opacity: '1', transform: 'translateY(0)' } },
      },
      maxWidth: {
        content: '1280px',
      },
    },
  },
  plugins: [],
};
